### ISSUE TITLE
Google/Bing locked SafeSearch by ISPs using CNAME hijacking

### ISSUE THREAD

**Issue by MasterKia** [uBlockOrigin/uBlock-issues#2167]
### Prerequisites

- [X] I verified that this is not a filter issue (MUST be reported at [filter issue tracker](https://github.com/uBlockOrigin/uAssets/issues))
- [X] This is not a [support issue or a question](https://old.reddit.com/r/uBlockOrigin/)
- [X] I performed a [cursory search of the issue tracker](https://github.com/uBlockOrigin/uBlock-issues/issues?q=is%3Aissue) to avoid opening a duplicate issue
- [X] The issue is not present after wholly disabling uBlock Origin ("uBO") in the browser
- [X] I checked the [documentation](https://github.com/gorhill/uBlock/wiki) to understand that the issue I report is not a normal behavior

### I tried to reproduce the issue when...

- [X] uBO is the only extension
- [X] uBO with default lists/settings
- [X] using a new, unmodified browser profile

### Description

ISPs in Iran are using SafeSearch CNAME to lock/force SafeSearch for _all_ users.
I use Firefox with `Uncloak canonical names` turned on, but I'm still experiencing this problem on google.com and bing.com.
Even `||forcesafesearch.google.com^$all` or `||strict.bing.com^$all` does nothing.

[Google Search Settings](https://www.google.com/preferences):
https://user-images.githubusercontent.com/17685483/178502343-d21efe1e-ab04-4436-9d38-3f27833004c6.png

[Bing Search Settings](https://www.bing.com/account/general):
https://user-images.githubusercontent.com/17685483/178502399-972fb43c-42af-4885-8334-86ebdfc4f717.png

I could only solve it using DoH in Firefox or DNS in Windows Network settings or using a VPN.

Related articles:
https://www.shahrsakhtafzar.com/fa/news/internet-operator/40486-google-safe-search-is-locked-in-iran
https://www.zoomit.ir/tech-iran/384198-google-safe-search-enabled-by-default/

[Google](https://support.google.com/websearch/answer/186669):
> To force SafeSearch for your network, you’ll need to update your DNS configuration. Set the DNS entry for www.google.com to be a **CNAME** for forcesafesearch.google.com.

[Bing](https://support.microsoft.com/en-us/topic/block-adult-content-with-safesearch-946059ed-992b-46a0-944a-28e8fb8f1814):
> To force SafeSearch for your network, you’ll need to update your DNS configuration. Set the DNS entry for [www.bing.com](http://www.bing.com/) to be a **CNAME** for strict.bing.com. You’ll want to use a **CNAME** rather than the strict.bing.com IP as the **CNAME** will continue to work even if the IP for strict.bing.com changes.

### A specific URL where the issue occurs

https://google.com

### Steps to Reproduce

1- Use Iran ISP
2- Go to https://www.google.com/preferences or https://www.bing.com/account/general.
3- I can't turn off SafeSearch.

### Expected behavior

I should be able to turn off SafeSearch.

### Actual behavior

I can't turn off SafeSearch.

### uBlock Origin version

1.43.0

### Browser name and version

Firefox 102.0.1

### Operating System and version

Windows 11

**Comment by u-RraaLL** [comment]
Have you tried changing to a not-automatic (your ISP's) DNS on your devices?

**Comment by MasterKia** [comment]
> I could only solve it using DoH in Firefox or DNS in Windows Network settings

Yes and it works.
However this involves CNAME and uBlock Origin has this "Uncloak canonical names" feature but it doesn't prevent this incident from happening.

**Comment by gwarser** [comment]
Even if you can block it by CNAME, then what? You will not be able to access the search page. uBO is not a DNS resolver.

**Comment by MasterKia** [comment]
> Not something related to uBO.

Related to "Uncloak canonical names".


> [**CNAME-uncloaking third-party trackers in Firefox**](https://en.wikipedia.org/wiki/UBlock_Origin#CNAME-uncloaking_third-party_trackers_in_Firefox)
It's a technique used by some sites to bypass third-party tracker blocking. 
These sites link to URLs that are sub-domains of the page's domain, **_but those sub-domains resolve to third-party hosts via a CNAME record_**.

This is exactly what's happening right now, the `google.com` domain resolves to `forcesafesearch.google.com` using CNAME.
> To force SafeSearch for your network, set the DNS entry for www.google.com to be a CNAME for forcesafesearch.google.com.

The only difference is that instead of a site, the ISP is doing this.
So uBO can only uncloak CNAME if sites are doing this but not when ISPs do it?
Maybe this could be fixed.


> Even if you can block it by CNAME, then what? You will not be able to access the search page.

I don't care about it being _blocked_, I _tried_ blocking it to see if uBO does really notice it happening and can stop it or not.


> uBO is not a DNS resolver.

I know but uBO does [DNS lookups](https://github.com/uBlockOrigin/uBlock-issues/issues/780#issuecomment-567270485).

**Comment by MasterKia** [comment]
TL;DR: It's established that uBO can uncloak CNAME when it's happening by a site, but why can't it also uncloak CNAME when it's happening by the ISP?

**Comment by MasterKia** [comment]
> you need to configure your browser

I'm afraid that's not possible on Firefox Mobile but it's possible on Firefox/Chrome desktop.
I guess on mobile I should configure DNS for the OS.

**Comment by MasterKia** [comment]
> bromite (though bromite is sometimes out of date) are good for ad blocking

Does the Bromite blocker support anything _other than_ network filters?
https://github.com/bromite/bromite/wiki/AdBlocking
> This engine does not currently support CSS rules; as a result some sites will show ads.

Android Private DNS doesn't work for me, maybe ISP somehow blocked DoT?

**Comment by MasterKia** [comment]
Yeah, after some research/asking it turns out that all ISPs have blocked the port 853 which is used by DoT on Android Private DNS.

So no DoT.

**Comment by u-RraaLL** [comment]
@gwarser should this maybe be converted into a discussion instead?

**Comment by MasterKia** [comment]
> You can use rethinkdns client if you want DoH for android

Thanks.








> I'm aware this is off-topic for uBlock Origin (uBO) so i'll probably stop now

Well the issue is already closed _anyway_.

**Comment by MasterKia** [comment]
Yes, people who've participated in the issue get notified and they could _unsubscribe_ from the issue if they want.

Also @gwarser might convert this "issue" into a "discussion".

**Comment by MasterKia** [comment]
No, but you'd have to self-host (because public instances mostly hit quota and get time-limited by Google)?

StartPage is blocked but DuckDuckGo isn't.
